### DISCUSSION TITLE
How do you debug tests?

### DISCUSSION THREAD

**Discussion by GP4cK** [TriPSs/nestjs-query#117] [Q&A] [answered]
I'm using VSCode and the [Jest Runner Extension](https://marketplace.visualstudio.com/items?itemName=firsttris.vscode-jest-runner). Usually, this adds `Run | Debug` on top of every `describe` / `it` and I can debug individual tests like this with breakpoints.

From what I read [here](https://github.com/nrwl/nx/issues/3992#issuecomment-719625701) and [there](https://blog.nrwl.io/vs-code-jest-improved-eslint-support-typescript-4-storybook-6-and-more-with-nx-10-3-faf7c12fe556) it should be possible but I can't even see the `Run | Debug`.

Do you mind sharing your `.vscode/launch.json` and `.vscode/settings.json`?

**Comment by TriPSs** [comment]
I do not use vscode so I'm not able to share it 😅 Maybe there is someone else that uses it that can share it.

**Reply by GP4cK** [reply]
For now I use this:

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "node",
      "request": "launch",
      "name": "Debug test",
      "program": "${workspaceFolder}/node_modules/jest/bin/jest",
      "args": ["--runInBand", "--collectCoverage=false", "--config=${workspaceFolder}/jest.preset.js", "${file}"],
      "console": "integratedTerminal",
      "internalConsoleOptions": "neverOpen"
    }
  ]
}
```

With this I can open a test file, go to the Run & Debug Tab and debug the entire test file.

**Reply by GP4cK** [reply]
Actually it might be an issue just on certain files because they are too big.
Ex: `packages/query-typeorm/__tests__/services/typeorm-query.service.spec.ts` doesn't show `Run | Debug` but `packages/query-graphql/__tests__/resolvers/relations/read-relation.resolver.spec.ts` does.

**Reply by GP4cK** [reply, marked as answer]
In fact, the issue is that the VSCode extension doesn't support decorator for function parameters at this time.
I've made a PR to fix it: https://github.com/firsttris/vscode-jest-runner/pull/308

If you can't wait, it's very easy to package the code and use your own version: https://code.visualstudio.com/api/working-with-extensions/publishing-extension

**Reply by GP4cK** [reply]
Just to add on to this, you can also run a single test in watch mode like this:

```
 nx run e2e --project=examples --testFile=examples/typeorm-soft-delete/e2e/todo-item.resolver.spec.ts --testNamePattern="should return deletedTodos and their subTasks" --watch
 ```
